fairlearn/metrics/_metric_frame.py

```diff
@@ -117,24 +129,156 @@ def __init__(self,
+
+        y_pred : List, pandas.Series, numpy.ndarray, pandas.DataFrame
+                The predictions.
+
+        sensitive_features : List, pandas.Series, dict of 1d arrays, numpy.ndarray,
+         pandas.DataFrame
+            The sensitive features which should be used to create the subgroups.
+            At least one sensitive feature must be provided.
+            All names (whether on pandas objects or dictionary keys) must be strings.
+            We also forbid DataFrames with column names of ``None``.
+            For cases where no names are provided we generate names ``sensitive_feature_[n]``.
+
+        control_features : List, pandas.Series, dict of 1d arrays, numpy.ndarray, DataFrame
+            Control features are similar to sensitive features, in that they
+            divide the input data into subgroups.
+            Unlike the sensitive features, aggregations are not performed
+            across the control features - for example, the ``overall`` property
+            will have one value for each subgroup in the control feature(s),
+            rather than a single value for the entire data set.
+            Control features can be specified similarly to the sensitive features.
+            However, their default names (if none can be identified in the
+            input values) are of the format ``control_feature_[n]``.
+
+            **Note** the types returned by members of the class vary based on whether
+            control features are present.
+
+        sample_params : dict
+            Parameters for the metric function(s). If there is only one metric function,
+            then this is a dictionary of strings and array-like objects, which are split
+            alongside the ``y_true`` and ``y_pred`` arrays, and passed to the metric function.
+            If there are multiple metric functions (passed as a dictionary), then this is
+            a nested dictionary, with the first set of string keys identifying the
+            metric function name, with the values being the string-to-array-like dictionaries.
+        """
+        if not self._streaming:
+            raise Exception("This MetricFrame does not support adding data.")
+        check_consistent_length(y_true, y_pred, sensitive_features)
+
         y_t = _convert_to_ndarray_and_squeeze(y_true)
         y_p = _convert_to_ndarray_and_squeeze(y_pred)
 
-        func_dict = self._process_functions(metric, sample_params)
+        self._y_true.append(y_t)
+        self._y_pred.append(y_p)
+        self._s_f.append(sensitive_features)
+        if self._c_f:
+            self._c_f.append(control_features)
+
+        if sample_params:
+            if self._s_p is None:
+                self._s_p = [sample_params]
+            elif isinstance(self._s_p, list):
+                self._s_p.append(sample_params)
+            else:
+                raise ValueError("MetricFrame was initialized with `sample_params` already set.")
```

Comment:
I don't understand this error. Aren't there two cases which should be being caught:
- User provides `sample_params` on the first call to this method, but then forgets on a subsequent call
- User _doesn't_ provide `sample_params` on the first call, but tries to on some later call

Reply:
There is also the case where the user supplies `sample_params` at initialization. For example, to provide `average='binary'` to `skm.f1_score`. Then, the user is not allowed to supply `sample_params` after.

Reply:
Surely `average='binary'` isn't a _sample_ parameter? By definition, sample parameters have a unique value for each sample - things like sample weights.